dozer/cogs/filter.py

```diff
@@ -0,0 +1,256 @@
+                await ctx.send(embed=embed)
+
+    def load_filters(self, guild_id):
+        with db.Session() as session:
+            results = session.query(WordFilter).filter_by(guild_id=guild_id, enabled=True).all()
+            self.filter_dict[guild_id] = {}
+            for filter in results:
+                self.filter_dict[guild_id][filter.id] = re.compile(filter.pattern, re.IGNORECASE)
+
+    async def check_filters(self, message):
+        if message.author.id == self.bot.user.id:
+            return
+        with db.Session() as session:
+            roles = session.query(WordFilterRoleWhitelist).filter_by(guild_id = message.guild.id).all()
+        whitelisted_ids = set(role.role_id for role in roles)
+        if any(x.id in whitelisted_ids for x in message.author.roles):
+            return
+        filters = {}
+        try:
+            filters = self.filter_dict[message.guild.id]
+        except KeyError:
+            self.load_filters(message.guild.id)
+            filters = self.filter_dict[message.guild.id]
+        for id, filter in filters.items():
+            if filter.search(message.content) is not None:
+                await message.channel.send("{}, Banned word detected!".format(message.author.mention), delete_after=5.0)
+                time = datetime.datetime.utcnow()
+                with db.Session() as session:
+                    infraction = WordFilterInfraction(member_id=message.author.id, filter_id=id,
+                                                      timestamp=time,
+                                                      message=message.content)
+                    session.add(infraction)
+                await message.delete()
+
+    """Event Handlers"""
+
+    async def on_message(self, message):
+        await self.check_filters(message)
+
+    async def on_message_edit(self, before, after):
+        await self.check_filters(after)
+
+    """Commands"""
+
+    @group(invoke_without_command=True)
+    @guild_only()
+    async def filter(self, ctx, advanced: bool=False):
+        """List and manage filtered words"""
+        with db.Session() as session:
+            results = session.query(WordFilter).filter_by(guild_id=ctx.guild.id, enabled=True).all()
+        if not results:
+            # TODO: Make this a embed
+            await ctx.send("No filters found for this server. Use `{}filter add <>` to add one.".format(
+                ctx.bot.command_prefix))
+            return
+
+        fmt = 'ID {0.id}: `{0.friendly_name}`'
+        if advanced:
+            fmt += ': Pattern: `{0.pattern}`'
+
+        filter_text = '\n'.join(map(fmt.format, results))
+
+        embed = discord.Embed()
+        embed.title = title = "Filters for {}".format(ctx.guild.name)
+        embed.add_field(name="Filters", value=filter_text)
+        await self.check_dm_filter(ctx, embed)
+
+    filter.example_usage = """`{prefix}filter add test` - Adds test as a filter.
+`{prefix}filter remove 1` - Removes filter 1
```

Comment:
The `example_usage` setter uses `inspect.cleandoc` to fix indentation, so this can be indented one level to be consistent with other commands.